nat.go

```diff
@@ -25,3 +29,216 @@ type NAT struct {
 	LogicalPort string
 	ExternalID  map[interface{}]interface{}
 }
+
+func (odbi *ovndb) rowToNat(uuid string) *NAT {
+	cacheNAT, ok := odbi.cache[tableNAT][uuid]
+	if !ok {
+		return nil
+	}
+
+	nat := &NAT{
+		UUID:       uuid,
+		Type:       cacheNAT.Fields["type"].(string),
+		ExternalIP: cacheNAT.Fields["external_ip"].(string),
+		LogicalIP:  cacheNAT.Fields["logical_ip"].(string),
+		ExternalID: cacheNAT.Fields["external_ids"].(libovsdb.OvsMap).GoMap,
+	}
+
+	if mac, ok := cacheNAT.Fields["external_mac"]; ok {
+		switch mac.(type) {
+		case libovsdb.UUID:
+			nat.ExternalMAC = mac.(libovsdb.UUID).GoUUID
+		case string:
+			nat.ExternalMAC = mac.(string)
+		}
+	}
+
+	if lip, ok := cacheNAT.Fields["logical_port"]; ok {
+		switch lip.(type) {
+		case libovsdb.UUID:
+			nat.LogicalIP = lip.(libovsdb.UUID).GoUUID
+		case string:
+			nat.LogicalIP = lip.(string)
+		}
+
+	}
+
+	return nat
+}
+
+func (odbi *ovndb) lrNatAddImp(ROUTER string, TYPE string, EXTERNAL_IP string, LOGICAL_IP string, external_ids map[string]string, LogicalPortAndExternalMac ...string) (*OvnCommand, error) {
+	nameUUID, err := newRowUUID()
+	if err != nil {
+		return nil, err
+	}
+	row := make(OVNRow)
+
+	row["external_ip"] = EXTERNAL_IP
+
+	row["logical_ip"] = LOGICAL_IP
+
+	switch TYPE {
+	case "snat":
+		row["type"] = TYPE
+	case "dnat":
+		row["type"] = TYPE
+	case "dnat_and_snat":
+		row["type"] = TYPE
+	default:
+		return nil, ErrorOption
+	}
+
+	if uuid := odbi.getRowUUID(tableNAT, row); len(uuid) > 0 {
+		return nil, ErrorExist
+	}
+
+	// The logical_port and  external_mac  are  only  accepted
+	// when  router  is  a  distributed  router  (rather than a gateway
+	// router) and type is dnat_and_snat.
+	if row["type"] == "dnat_and_snat" {
+		switch len(LogicalPortAndExternalMac) {
+		case 0:
```

Comment:
best to use case 0, 2 ?

Reply:
The logical_port and  external_mac  are  only  accepted  when  router  is  a  distributed  router [ovn-nbctl](https://www.openvswitch.org/support/dist-docs/ovn-nbctl.8.html)  
```bash
root@esixCell:/home/jacky# ovn-nbctl lr-nat-add lrtest1 dnat_and_snat 172.23.23.43 172.2.32.244
root@esixCell:/home/jacky# ovn-nbctl lr-nat-add lrtest1 dnat_and_snat 172.23.23.43 172.2.32.244 br-int
ovn-nbctl: lr-nat-add with logical_port must also specify external_mac.
```

Reply:
case 0 it will out of range....

Reply:
ok thanks